crypto/bio/bio_lib.c

```diff
@@ -12,23 +12,63 @@
 #include <openssl/crypto.h>
 #include "bio_lcl.h"
 #include "internal/cryptlib.h"
-#include <openssl/stack.h>
+
+static BIO nilb = {
+    NULL, NULL, NULL,   /* method, callback, cb_arg */
+    0, 1, 0,            /* init, shutdown, flags */
+    0, 0,               /* retry_reason, num */
+    NULL, NULL, NULL,   /* ptr, next_bio, prev_bio */
+    1,                  /* references */
+    0L, 0L,             /* num_read, num_write */
+    { NULL },           /* ex_data */
+    NULL                /* lock */
+};
 
 BIO *BIO_new(const BIO_METHOD *method)
 {
-    BIO *ret = OPENSSL_malloc(sizeof(*ret));
+    BIO *bio = OPENSSL_malloc(sizeof(*bio));
 
-    if (ret == NULL) {
+    if (bio == NULL) {
         BIOerr(BIO_F_BIO_NEW, ERR_R_MALLOC_FAILURE);
         return (NULL);
     }
-    if (!BIO_set(ret, method)) {
-        OPENSSL_free(ret);
-        ret = NULL;
+
+    /* direct structure assignment */
+    *bio = nilb;
+
+    bio->lock = CRYPTO_THREAD_lock_new();
```

Comment:
what API uses the per-object lock?  we haven't made the promise of doing this, not sure we want to do it for just this one type this late in the release cycle.